espressostreamer/espresso_streamer.go

```diff
@@ -98,36 +95,106 @@ func NewEspressoStreamer(
-	defer s.messageMutex.Unlock()
 	s.currentMessagePos = currentMessagePos
 	s.nextHotshotBlockNum = currentHostshotBlock
 	s.messageWithMetadataAndPos = []*MessageWithMetadataAndPos{}
 }
 
 func (s *EspressoStreamer) Next() (*MessageWithMetadataAndPos, error) {
-	s.messageMutex.Lock()
-	defer s.messageMutex.Unlock()
+	result, err := s.Peek()
+	if err != nil {
+		return nil, err
+	}
 
-	message, found := FilterAndFind(&s.messageWithMetadataAndPos, func(msg *MessageWithMetadataAndPos) int {
+	// Advance the current message position, so that the next call to
+	// `Peek` or `Next` will return the next message
+	s.Advance()
+	return result, nil
+}
+
+func (s *EspressoStreamer) Peek() (*MessageWithMetadataAndPos, error) {
+	compareMessageWithCurrentPos := func(msg *MessageWithMetadataAndPos) int {
 		if msg.Pos == s.currentMessagePos {
-			return 0
+			return FilterAndFind_Target
 		}
 		if msg.Pos < s.currentMessagePos {
-			return -1
+			return FilterAndFind_Remove
 		}
-		return 1
-	})
-	if !found {
-		return nil, nil
+		return FilterAndFind_Keep
 	}
-	if message == nil {
-		// This should never happen.
-		return nil, fmt.Errorf("message is nil, but found is true")
+
+	messageIndex := FilterAndFind(&s.messageWithMetadataAndPos, compareMessageWithCurrentPos)
+
+	if messageIndex >= 0 {
+		return s.messageWithMetadataAndPos[messageIndex], nil
+	}
+
+	condition := func(messages []*MessageWithMetadataAndPos) bool {
+		for _, message := range messages {
+			if message.Pos == s.currentMessagePos {
+				return true
+			}
+		}
+		return false
 	}
 
+	err := s.QueueMessagesFromHotShotUntil(context.Background(), s.parseEspressoTransaction, condition)
+	if err != nil {
+		return nil, err
+	}
+
+	messageIndex = FilterAndFind(&s.messageWithMetadataAndPos, compareMessageWithCurrentPos)
+
+	if messageIndex >= 0 {
+		return s.messageWithMetadataAndPos[messageIndex], nil
+	}
+	return nil, fmt.Errorf("message not found, please check the condition")
+}
+
+// Call this function to advance the streamer to the next message
+func (s *EspressoStreamer) Advance() {
 	s.currentMessagePos += 1
-	return message, nil
+}
+
+// This function keep fetching hotshot blocks and parsing them until the condition is met.
+// It is a do-while loop, which means it will always execute at least once.
+//
+// Expose the *parseHotShotPayloadFn* to the caller for testing purposes
+func (s *EspressoStreamer) QueueMessagesFromHotShotUntil(
+	ctx context.Context,
+	parseHotShotPayloadFn func(tx espressoTypes.Bytes) ([]*MessageWithMetadataAndPos, error),
+	condition func(messages []*MessageWithMetadataAndPos) bool,
+) error {
+	var e error
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		messages, err := fetchNextHotshotBlock(ctx, s.espressoClient, s.nextHotshotBlockNum, parseHotShotPayloadFn, s.namespace)
+		if err != nil {
+			e = fmt.Errorf("%w: %w", FailedToFetchTransactionsErr, err)
+			// TODO: handle the case more appropriately
+			if condition(messages) || ctx.Err() != nil {
```

Comment:
should we add a timeout here?

Reply:
Yeah I think we should do some similar things. If timeout, then we can set a timeout context